src/material/badge/badge.ts

```diff
@@ -156,6 +156,20 @@ export class MatBadge extends _MatBadgeBase implements OnInit, OnDestroy, CanDis
       if (nativeElement.nodeType !== nativeElement.ELEMENT_NODE) {
         throw Error('matBadge must be attached to an element node.');
       }
+
+      const matIconTagName: string = 'mat-icon';
+
+      // Heads-up for developers to avoid putting matBadge on <mat-icon>
+      // as it is aria-hidden by default docs mention this at:
+      // https://material.angular.io/components/badge/overview#accessibility
+      if (
+        nativeElement.tagName.toLowerCase() === matIconTagName &&
+        nativeElement.getAttribute('aria-hidden') === 'true'
+      ) {
+        console.warn(
+          `Warning: detected matBadge on a "<mat-icon>" with "aria-hidden="true"\n${nativeElement.outerHTML}`,
```

Comment:
I think the warning should mention why this is a problem and what the user should do. Otherwise it's unclear why having a badge there is a problem.

Reply:
something like:
```
Warning: detected matBadge on "<mat-icon>" as it's aria-hidden="true" by default, making badge content invisible to screen readers. Consider adding aria-hidden="false" to "<mat-icon>" so matBadge description stay preserved to screen readers. 

${nativeElement.outerHTML}
```

idk, it took me 25m to come up with the above message. I'm not really good with English & explaining things, feel free to correct me on this.

Reply:
Maybe something like this?

> Detected a matBadge on an `aria-hidden` `<mat-icon>`. Consider setting `aria-hidden="false"` in order to surface the information assistive technology.